src/internal/lokiutil/testloki/logger.go

```diff
@@ -0,0 +1,138 @@
+package testloki
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/pachyderm/pachyderm/v2/src/internal/errors"
+	"github.com/pachyderm/pachyderm/v2/src/internal/lokiutil/client"
+	"github.com/pachyderm/pachyderm/v2/src/internal/pachconfig"
+	"github.com/pachyderm/pachyderm/v2/src/internal/pachd"
+	"github.com/pachyderm/pachyderm/v2/src/internal/pctx"
+	"github.com/pachyderm/pachyderm/v2/src/internal/randutil"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+)
+
+func WithTestLoki(l *TestLoki) pachd.TestPachdOption {
+	return pachd.TestPachdOption{
+		MutateEnv: func(env *pachd.Env) {
+			env.GetLokiClient = func() (*client.Client, error) {
+				return l.Client, nil
+			}
+		},
+		MutateConfig: func(config *pachconfig.PachdFullConfiguration) {
+			config.LokiLogging = true
+		},
+		MutateContext: func(ctx context.Context) context.Context {
+			templateHash := randutil.UniqueString("")[0:10]
+			procHash := randutil.UniqueString("")[0:5]
+			return pctx.Child(ctx, "", l.WithLoki(ctx, map[string]string{
+				"host":              "localhost",
+				"app":               "pachd",
+				"container":         "pachd",
+				"node_name":         "localhost",
+				"pod":               fmt.Sprintf("pachd-%v-%v", templateHash, procHash),
+				"pod_template_hash": templateHash,
+				"stream":            "stderr",
+				"suite":             "pachyderm",
+			}))
+		},
+	}
+}
+
+// newZapCore returns a zapcore.Core that sends logs to this Loki instance.
+func (l *TestLoki) newZapCore(ctx context.Context) *lokiCore {
+	return &lokiCore{
+		ctx: ctx,
+		l:   l,
+		enc: zapcore.NewJSONEncoder(zapcore.EncoderConfig{
+			TimeKey:        "time",
+			EncodeTime:     zapcore.RFC3339NanoTimeEncoder,
+			LevelKey:       "severity",
+			EncodeLevel:    zapcore.LowercaseLevelEncoder,
+			MessageKey:     "message",
+			NameKey:        "logger",
+			CallerKey:      "caller",
+			FunctionKey:    zapcore.OmitKey,
+			StacktraceKey:  "stacktrace",
+			LineEnding:     zapcore.DefaultLineEnding,
+			EncodeDuration: zapcore.SecondsDurationEncoder,
+			EncodeCaller:   zapcore.ShortCallerEncoder,
+		}),
+		labels: map[string]string{"suite": "pachyderm"},
+	}
+}
+
+// WithLoki returns a pctx context option that will send logs to this Loki instance.
+func (l *TestLoki) WithLoki(sendCtx context.Context, lokiLabels map[string]string) pctx.Option {
+	lc := l.newZapCore(sendCtx)
+	lc.labels = lokiLabels
+	return pctx.WithOptions(zap.WrapCore(func(c zapcore.Core) zapcore.Core {
+		return zapcore.NewTee(c, lc)
+	}))
+}
+
+// lokiCore sends JSON log messages to the provided TestLoki instance.  If you were going to use
+// this code in production, you would want to skip the JSON serialization step, buffer lines to send
+// them in batches, and read the time from the Entry instead of using time.Now().  (We use
+// time.Now() here because it's more like what promtail does; reads the line and adds its own
+// timestamp.)
+type lokiCore struct {
+	ctx    context.Context
```

Comment:
You really don't want logging to block, that's why it doesn't take a context.  If we were doing this for real, we'd write to a buffered channel and have a background process take logs out of the channel and send to loki.  Then a context wouldn't be needed in the logger, only in the background sender.  But for tests, the synchronization is a benefit; if you call an RPC foo and it returns, you'll expect Loki to have the logs for that request.  Since it's synchronous, that happens.  But it's not good for performance in general.